KAFKA-14133: Replace EasyMock with Mockito in streams test

mdedetrich opened the pull request:
Batch 3 of the tests detailed in https://issues.apache.org/jira/browse/KAFKA-14133 which use EasyMock and need to be moved to Mockito.

EDIT: This issue has been solved
Unfortunately some tests are failing and I am not exactly sure how and why this is happening, specifically

* `shouldNotUpdateLimitForNonSourceStandbyChangelog`
* `shouldOnlyRestoreStandbyChangelogInUpdateStandbyState`
* `shouldRestoreToLimitInStandbyState`

As you can see from the diff in this PR, the changes from `EasyMock` to `Mockito` in these functions was quite trivial/mechanical but the tests are failing with errors that are confusing to me, i.e.

```
[2022-08-17 12:13:33,803] DEBUG test-reader Transiting to update standby tasks: {} (org.apache.kafka.streams.processor.internals.StoreChangelogReader:317)
[2022-08-17 12:13:33,803] DEBUG test-reader Resumed partitions [] from the restore consumer (org.apache.kafka.streams.processor.internals.StoreChangelogReader:899)
[2022-08-17 12:13:33,804] DEBUG test-reader Added partitions [topic-0] to the restore consumer, current assignment is [topic-0] (org.apache.kafka.streams.processor.internals.StoreChangelogReader:861)
[2022-08-17 12:13:33,804] DEBUG test-reader Start restoring changelog partition topic-0 from current offset 0 to end offset 7. (org.apache.kafka.streams.processor.internals.StoreChangelogReader:918)
[2022-08-17 12:13:33,805] WARN test-reader Encountered org.apache.kafka.clients.consumer.OffsetOutOfRangeException fetching records from restore consumer for partitions [topic-0], it is likely that the consumer's position has fallen out of the topic partition offset range because the topic was truncated or compacted on the broker, marking the corresponding tasks as corrupted and re-initializing it later. (org.apache.kafka.streams.processor.internals.StoreChangelogReader:438)
org.apache.kafka.clients.consumer.OffsetOutOfRangeException: Offsets out of range with no configured reset policy for partitions: {topic-0=1}
	at org.apache.kafka.clients.consumer.MockConsumer.poll(MockConsumer.java:208)
	at org.apache.kafka.streams.processor.internals.StoreChangelogReader.restore(StoreChangelogReader.java:433)
	at org.apache.kafka.streams.processor.internals.StoreChangelogReaderTest.shouldRestoreToLimitInStandbyState(StoreChangelogReaderTest.java:836)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.mockito.internal.junit.JUnitSessionStore$1.evaluateSafely(JUnitSessionStore.java:55)
	at org.mockito.internal.junit.JUnitSessionStore$1.evaluate(JUnitSessionStore.java:43)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runners.Suite.runChild(Suite.java:128)
	at org.junit.runners.Suite.runChild(Suite.java:27)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at org.junit.vintage.engine.execution.RunnerExecutor.execute(RunnerExecutor.java:42)
	at org.junit.vintage.engine.VintageTestEngine.executeAllChildren(VintageTestEngine.java:80)
	at org.junit.vintage.engine.VintageTestEngine.execute(VintageTestEngine.java:72)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:53)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:99)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:79)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:75)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy2.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$3.run(TestWorker.java:193)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:133)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)

Tasks [null] are corrupted and hence needs to be re-initialized
org.apache.kafka.streams.errors.TaskCorruptedException: Tasks [null] are corrupted and hence needs to be re-initialized
	at app//org.apache.kafka.streams.processor.internals.StoreChangelogReader.restore(StoreChangelogReader.java:446)
	at app//org.apache.kafka.streams.processor.internals.StoreChangelogReaderTest.shouldRestoreToLimitInStandbyState(StoreChangelogReaderTest.java:836)
```

@cadonna @clolov @C0urante @ijuma Maybe you have an idea what is going on here? Also note that due to Mockito not working with inline `thenReturn` mock statements (see https://github.com/mockito/mockito/wiki/FAQ#can-i-thenreturn-an-inlined-mock-) I did some refactorings in `StoreChangelogReaderTest`, i.e. `mockTasks` and `mockTask` have been moved from inner local methods to the class. This was also done as an attempt to solve the aforementioned problems but it ended up having zero effect (also note in general apart from these problems it had zero effect on the tests).

Also note that `@SuppressWarnings("unchecked")` is necessary because unlike EasyMock, Mockito doesn't have any support for generics in the `mock` method (see https://github.com/mockito/mockito/issues/1531)

Finally with `ProcesserStateManagerTest` some verifications, stubbed calls have been removed for the same reasons behind https://github.com/apache/kafka/pull/12505#discussion_r947652273, i.e. `MockitoJUnitRunner.StrictStubs` was complaining that there was extra erroneous stubs/verifications and I needed to remove them to make `StrictStubs` happy. This stubbing was also unnecessary in `EasyMock`.

mdedetrich committed:
Replace EasyMock with Mockito in streams test

mdedetrich commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamThreadTest.java:
```diff
@@ -2912,34 +2872,34 @@ void runOnce() {
+        when(taskManager.allTasks())
+            .thenReturn(Collections.singletonMap(taskId, runningTask));
+        when(taskManager.commit(Collections.singleton(runningTask))).thenReturn(1);
+        for (int i = 0; i < numberOfPurges; i++) {
```
This for loop was added to actually simulate the number of purges that I assume the `mockTaskManagerPurge` is trying to emulate. Not exactly sure how `EasyMock` did this but this version is more explicit and clear.

C0urante commented:
Hey @mdedetrich--it looks like the issue here is caused by differences in default return values for boxed primitive types between Mockito mocks and EasyMock nice mocks; the former will return 0, but the latter will return null.

Adding `when(storeMetadata.offset()).thenReturn(null);` to the beginning of each of the failing test cases mentioned in the description seems to fix the issue.

I'm not too familiar with the Streams codebase, so probably best to get review from someone else, but if there are any other strange issues that don't seem Streams-specific, let me know!

mdedetrich commented:
@C0urante Thanks for finding the root cause, will definitely keep this as a mental note when doing other migrations. I'll update the PR later tonight

mdedetrich commented:
PR has been rebased and pushed which fixes the problem described earlier, i.e. usage of `when(storeMetadata.offset()).thenReturn(null);`

mdedetrich commented:
@cadonna I have just rebased the PR with the latest streams changes that got merged into. trunk. I also did some more improvements for `StreamThreadTest` with the usage of `MockitoJUnitRunner.StrictStubs`, i.e. before where we had

```java
verify(taskManager)
```

we now have

```java
verify(taskManager).handleCorruption(corruptedTasks);
```

Also `MockitoJUnitRunner.StrictStubs` revealed pointless stubs have been removed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java:
```diff
@@ -308,29 +300,21 @@ public void shouldRestoreTimestampedStoreWithConverter() {
         reset(store);
-        expect(store.name()).andStubReturn(persistentStoreName);
-        store.close();
-        replay(store);
+        when(store.name()).thenReturn(persistentStoreName);
```
Is the `reset()` and the stubbing of `store.name()` still needed? The stubbing is the same as on line 304. Since we are not doing a new replay, I am not sure that we still need the `reset()` with Mockito. You think you can moves all `verify()` to the end of the method.

mdedetrich replied:
No its not, I just removed this along with the `verify(store).close();` since its pointless

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java:
```diff
@@ -360,12 +339,11 @@ public void shouldRecycleStoreAndReregisterChangelog() {
 
         reset(context, store);
         context.uninitialize();
```
I think you should remove this line. It was a verification in EasyMock and now with Mockito it is a call on the mock, which does not make a lot of sense. You make the call and in the next line you immediately verify that you made the call. Instead you should verify the call that is done within `stateMgr.registerStateStores(singletonList(store), context)`. Additionally to remove `context.uninitialize();` you need to move `verify(context).uninitialize();` below `stateMgr.registerStateStores(singletonList(store), context)`.

mdedetrich replied:
Done.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StandbyTaskTest.java:
```diff
@@ -177,10 +181,7 @@ public void shouldThrowLockExceptionIfFailedToLockStateDirectory() throws IOExce
-        stateManager.registerStateStores(EasyMock.anyObject(), EasyMock.anyObject());
-        EasyMock.expect(stateManager.changelogPartitions()).andReturn(Collections.emptySet()).anyTimes();
-        EasyMock.replay(stateManager);
+        stateManager.registerStateStores(any(), any());
```
This is also a call on the mock which makes the verification on line 199 always true, independently of the call under test.

mdedetrich replied:
Done.

cadonna replied:
I meant that you should remove the call on the mock, because mocks should not be directly called in the test code. Mocks should be called within the call under test. In this case this is `task.initializeIfNeeded()`. Here it is really important that `stateManager.registerStateStores()` is only called once, since a second call to `task.initializeIfNeeded()` should not register the state stores again. The test should be like this:
```
    public void shouldTransitToRunningAfterInitialization() {
        task = createStandbyTask();

        assertEquals(CREATED, task.state());

        task.initializeIfNeeded();

        assertEquals(RUNNING, task.state());

        // initialize should be idempotent
        task.initializeIfNeeded();

        assertEquals(RUNNING, task.state());
        verify(stateManager).registerStateStores(any(), any());
    }
```
You are right to say that the stubbings were unnecessary here since they returned what is the default in Mockito.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StandbyTaskTest.java:
```diff
@@ -195,12 +196,11 @@ public void shouldTransitToRunningAfterInitialization() {
         assertEquals(RUNNING, task.state());
 
-        EasyMock.verify(stateManager);
+        verify(stateManager, atLeastOnce()).registerStateStores(any(), any());
```
```suggestion
        verify(stateManager).registerStateStores(any(), any());
```

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StandbyTaskTest.java:
```diff
@@ -212,31 +212,29 @@ public void shouldThrowIfCommittingOnIllegalState() {
     @Test
     public void shouldAlwaysCheckpointStateIfEnforced() {
         stateManager.flush();
-        EasyMock.expectLastCall().once();
+        verify(stateManager, times(1)).flush();
```
Again, you make a call on the mock and verify that you made the call. In EasyMock you need to make the calls on the mock to specify what calls you expect. The call to `replay()` ends the specification phase. When you `verify()`, you verify that the call under test indeed called to methods you specified. 
This is not the case in Mockito. In Mockito the mocks record the calls and you verify what was called. Making calls on the mock directly in the test method does not make sense. The methods of the mocks should be called within the call under test.

mdedetrich replied:
Done.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StandbyTaskTest.java:
```diff
@@ -212,31 +212,29 @@ public void shouldThrowIfCommittingOnIllegalState() {
-        EasyMock.expectLastCall().once();
-        EasyMock.expect(stateManager.changelogOffsets()).andReturn(Collections.emptyMap()).anyTimes();
-        EasyMock.replay(stateManager);
+        verify(stateManager, times(1)).checkpoint();
+        when(stateManager.changelogOffsets()).thenReturn(Collections.emptyMap());
```
Again here.

mdedetrich replied:
Done

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StandbyTaskTest.java:
```diff
@@ -212,31 +212,29 @@ public void shouldThrowIfCommittingOnIllegalState() {
         task.maybeCheckpoint(true);
 
-        EasyMock.verify(stateManager);
+        verify(stateManager, atLeastOnce()).flush();
```
```suggestion
        verify(stateManager).flush();
```

mdedetrich replied:
Done

cadonna reviewed (commented):
@mdedetrich Thank you for the PR!

I started to review the PR. Please see my comments about how to use EasyMock and Mockito. Please adapt the PR also in the other parts that show similar issues.

mdedetrich commented:
Thanks for the comments, I am still get my head around the tests and what precisely is being tested.

I forgot to mention earlier that `StreamTaskTest` needs more work (doing this now, will push changes). There are other fixes needed due to the `MockitoJUnitRunner.StrictStubs`, i.e. some tests throw exceptions and expect that exception path to never be hit and this needs to be replaced with `verify(times(0))` or something similar.

mdedetrich commented:
@cadonna  Okay so I have just pushed and rebased the PR with these changes

* As you have rightly pointed out there were a lot of things being stubbed/tested which made no sense due to peculiarities of how `EasyMock` works (which I am still trying to get my head around). I believe I have removed most of the unnecessary `reset`s/stubs. There are still some `verify`'s which can be argued are pointless (mainly in the setup of the mock), if you still want to remove these then let me know
* `StreamTaskTest` previously with `EasyMock` used stub exceptions to verify a method was not meant to be called, i.e.
  ```java
  EasyMock.expectLastCall().andThrow(new AssertionError("Should not have tried to checkpoint"));
  ```
  This has now been replaced with using `never`, i.e. the above example translates to `verify(stateManager, never()).checkpoint();` which

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java:
```diff
@@ -310,48 +301,30 @@ public void shouldRestoreTimestampedStoreWithConverter() {
 
-        reset(store);
-        expect(store.name()).andStubReturn(persistentStoreName);
-        store.close();
```
You should account for this by putting `verify(store).close()` after line 314. Or am I missing something?

mdedetrich replied:
Adding `verify(store).close()` at this point causes the test to fail, `store.close()` is not executed at this point.

cadonna replied:
Then I do not understand why the test passed before. I suppose there is a mistake in the test setup.

mdedetrich replied:
So if I use the `Mockito.reset` to simulate what the original `EasyMock` was doing, i.e.

```java
reset(store);
when(store.name()).thenReturn(persistentStoreName);
store.close();
```

The the test passes. Note that you told me to remove this beforehand (see https://github.com/apache/kafka/pull/12524#discussion_r968477415)

cadonna replied:
I think we are misunderstanding each other. With after line 314, I meant after `stateMgr.close();`. It could be that I mixed up the line number. Currently the test does not pass (as you can see in the builds) because the verification `verify(store).close();` is before `stateMgr.close();`.
The test should look like this:
```
    public void shouldUnregisterChangelogsDuringClose() {
        final ProcessorStateManager stateMgr = getStateManager(Task.TaskType.ACTIVE);
        final StateStore store = mock(StateStore.class);
        when(store.name()).thenReturn(persistentStoreName);

        stateMgr.registerStateStores(singletonList(store), context);
        verify(context).uninitialize();
        verify(store).init((StateStoreContext) context, store);

        stateMgr.registerStore(store, noopStateRestoreCallback, null);
        assertTrue(changelogReader.isPartitionRegistered(persistentStorePartition));

        stateMgr.close();
        verify(store).close();
        assertFalse(changelogReader.isPartitionRegistered(persistentStorePartition));
    }
```

mdedetrich replied:
Ah that makes much more sense, done and committed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorStateManagerTest.java:
```diff
@@ -360,14 +333,8 @@ public void shouldRecycleStoreAndReregisterChangelog() {
         assertThat(stateMgr.getStore(persistentStoreName), equalTo(store));
 
-        reset(context, store);
-        context.uninitialize();
```
To migrate all verification, you also need to migrate this to `verify(context).uninitialize()` after line 336. Note here, I am not sure if you also need to count the previous call to `context.uninitialize()` verified on line 326. That is, you might need to add `verify(context, times(2)).uninitialize()`.

mdedetrich replied:
So counts are absolute which means that modifying `verify(context).uninitialize();` to `verify(context, times(2)).uninitialize()` is not necessary.  I did add `verify(context).uninitialize();` on line 336.

cadonna replied:
The pre-migration code verfies that `context.uninitialize()` is called during the second `stateMgr.registerStateStores(singletonList(store), context);`. That means `verify(context).uninitialize();`  should be after the second `stateMgr.registerStateStores(singletonList(store), context);`. There you need to use `verify(context, times(2)).uninitialize()` since the test fails otherwise. The test should look like this:

```
    public void shouldRecycleStoreAndReregisterChangelog() {
        final ProcessorStateManager stateMgr = getStateManager(Task.TaskType.ACTIVE);
        final StateStore store = mock(StateStore.class);
        when(store.name()).thenReturn(persistentStoreName);
        stateMgr.registerStateStores(singletonList(store), context);

        verify(context).uninitialize();
        verify(store).init((StateStoreContext) context, store);

        stateMgr.registerStore(store, noopStateRestoreCallback, null);
        assertTrue(changelogReader.isPartitionRegistered(persistentStorePartition));

        stateMgr.recycle();
        assertFalse(changelogReader.isPartitionRegistered(persistentStorePartition));
        assertThat(stateMgr.getStore(persistentStoreName), equalTo(store));

        stateMgr.registerStateStores(singletonList(store), context);

        verify(context, times(2)).uninitialize();
        assertTrue(changelogReader.isPartitionRegistered(persistentStorePartition));
    }
```

cadonna replied:
Maybe also here we misunderstood each other regarding the line numbers.

mdedetrich replied:
That makes a lot more sense, just committed your change and pushed.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StandbyTaskTest.java:
```diff
@@ -177,11 +179,6 @@ public void shouldThrowLockExceptionIfFailedToLockStateDirectory() throws IOExce
     @Test
     public void shouldTransitToRunningAfterInitialization() {
-        EasyMock.expect(stateManager.changelogOffsets()).andStubReturn(Collections.emptyMap());
-        stateManager.registerStateStores(EasyMock.anyObject(), EasyMock.anyObject());
```
I think there are still some misunderstandings how EasyMock works. When you call a method on the mock, this call is registered in the mock. Registration means that the method is expected to be called during the call of the method under test. When you call `EasyMock.replay(mock)` you end the registration phase. From now on the mock stores the methods actually called on it and returns the return values specified on the stubbings (i.e., `expect(mock).andReturn(...)`) when the stubbed method is called. After replaying the mock, you call the method under test. Finally, with `EasyMock.verify(mock)` you verify that only the registered method calls were performed during the call to the method under test.
In this specific case, the call `registerStateStores(EasyMock.anyObject(), EasyMock.anyObject())` was registered on the mock `stateManager` as an expected call. With `EasyMock.verify(stateManager);` it is verified that method `registerStateStores()` was called on the mock.
In your migration this verification disappeared.

mdedetrich replied:
Yes this part of `EasyMock` I understand, the problem is that a lot of the tests had stubbings that were unnecessary which was being picked up being `MockitoJUnitRunner.StrictStubs.class`. Ontop of this `Easymock.verify(stateManager)` has no equivalent in mockito since its considered bad practice to just verify any interaction on a class (Mockito forces you to specify which method is being called, not just `Easymock.verify(someClass)`. Since the `Easymock.verify(someClass)` was typically placed at the bottom of the test methods so it was hard to tell what exactly was expected to mock and in some cases it was pointless.

Evidently I did some mistakes so I have put back `stateManager.registerStateStores(any(), any());` however note that `stateManager.registerStateStores()` is being called twice so I used `atLeastOnce`.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StandbyTaskTest.java:
```diff
@@ -211,32 +205,22 @@ public void shouldThrowIfCommittingOnIllegalState() {
 
     @Test
     public void shouldAlwaysCheckpointStateIfEnforced() {
-        stateManager.flush();
```
You missed to verify this call.

mdedetrich replied:
Done.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StandbyTaskTest.java:
```diff
@@ -211,32 +205,22 @@ public void shouldThrowIfCommittingOnIllegalState() {
 
     @Test
     public void shouldOnlyCheckpointStateWithBigAdvanceIfNotEnforced() {
-        stateManager.flush();
```
You missed to verify this call.

mdedetrich replied:
Done

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StandbyTaskTest.java:
```diff
@@ -211,32 +205,22 @@ public void shouldThrowIfCommittingOnIllegalState() {
     public void shouldOnlyCheckpointStateWithBigAdvanceIfNotEnforced() {
-        stateManager.flush();
-        EasyMock.expectLastCall().once();
-        stateManager.checkpoint();
```
You missed to verify this call.

cadonna replied:
This call still needs to be verified with `verify(stateManager).checkpoint()` after `verify(stateManager).flush()`.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StandbyTaskTest.java:
```diff
@@ -247,23 +231,19 @@ public void shouldOnlyCheckpointStateWithBigAdvanceIfNotEnforced() {
+        when(stateManager.changelogOffsets()).thenReturn(Collections.emptyMap());
         stateManager.flush();
-        EasyMock.expectLastCall();
+        verify(stateManager, atLeastOnce()).flush();
```
You again call a function on the mock and immediately verify it. Maybe you just missed to remove it here, since you added the verification on line 259. However, I do not understand why you use `atLeastOnce()`. It should be called only once according to the pre-migration code. If you remove `stateManager.flush()` on line 239, you will see that it is indeed only called once.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StandbyTaskTest.java:
```diff
@@ -247,23 +231,19 @@ public void shouldOnlyCheckpointStateWithBigAdvanceIfNotEnforced() {
-                .andReturn(Collections.singletonMap(partition, 11000L));
-        EasyMock.expect(stateManager.changelogPartitions()).andReturn(Collections.singleton(partition)).anyTimes();
-        EasyMock.replay(stateManager);
+        verify(stateManager, times(1)).checkpoint();
```
This verification should be moved to the end of the test. You do not need to use `times(1)` that is the default behavior. Do not forget to remove `stateManager.checkpoint();` on line 241. It does not make sense to call methods on the mocks in Mockito. Only in EasyMock calling methods on the mock is a way to specify the expected behavior.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StandbyTaskTest.java:
```diff
@@ -276,32 +256,21 @@ public void shouldFlushAndCheckpointStateManagerOnCommit() {
-        EasyMock.expect(stateManager.changelogOffsets()).andStubReturn(Collections.emptyMap());
-        stateManager.close();
-        EasyMock.expectLastCall().andThrow(new ProcessorStateException("KABOOM!")).anyTimes();
-        stateManager.flush();
```
You missed to verify this call at the end of the test with `verify(stateManager, never()).flush()`.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StandbyTaskTest.java:
```diff
@@ -276,32 +256,21 @@ public void shouldFlushAndCheckpointStateManagerOnCommit() {
-        EasyMock.expectLastCall().andThrow(new ProcessorStateException("KABOOM!")).anyTimes();
-        stateManager.flush();
-        EasyMock.expectLastCall().andThrow(new AssertionError("Flush should not be called")).anyTimes();
-        stateManager.checkpoint();
```
You missed to verify this call at the end of the test with `verify(stateManager, never()).checkpoint()`.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StandbyTaskTest.java:
```diff
@@ -313,36 +282,29 @@ public void shouldNotFlushAndThrowOnCloseDirty() {
         task.closeDirty();
 
-        EasyMock.verify(stateManager);
+        verify(stateManager, atLeastOnce()).close();
```
Why this verification?

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StandbyTaskTest.java:
```diff
@@ -313,36 +282,29 @@ public void shouldNotFlushAndThrowOnCloseDirty() {
-                .andReturn(Collections.singletonMap(partition, 60L));
-        EasyMock.expect(stateManager.changelogPartitions()).andReturn(Collections.singleton(partition)).anyTimes();
-        EasyMock.replay(stateManager);
+        verify(stateManager, times(1)).checkpoint();
```
Please move this verification to after the call under test.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StandbyTaskTest.java:
```diff
@@ -313,36 +282,29 @@ public void shouldNotFlushAndThrowOnCloseDirty() {
         stateManager.close();
-        EasyMock.expectLastCall();
+        verify(stateManager, atLeastOnce()).close();
         stateManager.checkpoint();
```
Remove this call on the mock. In Mockito it does not make sense.

mdedetrich replied:
Done

cadonna replied:
The call on the mock is still there.

mdedetrich replied:
Are you checking the latest state of the PR and pointing out `verify(stateManager, atLeastOnce()).close();`? There is a `verify(stateManager, atLeastOnce()).close();` but its at the end of the test and its now ordered, this is the current latest state

```java
@Test
public void shouldSuspendAndCommitBeforeCloseClean() {
    when(stateManager.changelogOffsets())
        .thenReturn(Collections.singletonMap(partition, 60L));
    final InOrder stateManagerOrdered = inOrder(stateManager);
    stateManager.close();
    stateManager.checkpoint();
    stateManagerOrdered.verify(stateManager, times(1)).checkpoint();
    when(stateManager.changelogOffsets())
            .thenReturn(Collections.singletonMap(partition, 60L));
    final MetricName metricName = setupCloseTaskMetric();

    task = createStandbyTask();
    task.initializeIfNeeded();
    task.suspend();
    task.prepareCommit();
    task.postCommit(true);
    task.closeClean();

    assertEquals(Task.State.CLOSED, task.state());

    final double expectedCloseTaskMetric = 1.0;
    verifyCloseTaskMetric(expectedCloseTaskMetric, streamsMetrics, metricName);

    verify(stateManager).changelogOffsets();
    stateManagerOrdered.verify(stateManager, atLeastOnce()).close();
}
```

cadonna replied:
Yes, I am looking at the latest commit. 
I added some inline comments to the code you posted.
```java
@Test
public void shouldSuspendAndCommitBeforeCloseClean() {
    when(stateManager.changelogOffsets())
        .thenReturn(Collections.singletonMap(partition, 60L));
    final InOrder stateManagerOrdered = inOrder(stateManager);
    stateManager.close(); // <- This is a call directly on the mock. 
     // With EasyMock that is the way to specify expected calls on a mock. 
     // With Mockito you must not do this. 
     // With Mockito you verify the expected calls at the end with Mockito.verify().
    stateManager.checkpoint(); // <- This is also a direct call on the mock. 
     // The same applies as above      
    stateManagerOrdered.verify(stateManager, times(1)).checkpoint(); // <- This is always true 
    // if you make the call stateManager.close() above.
    when(stateManager.changelogOffsets())
            .thenReturn(Collections.singletonMap(partition, 60L));
    final MetricName metricName = setupCloseTaskMetric();

    task = createStandbyTask();
    task.initializeIfNeeded();
    task.suspend();
    task.prepareCommit();
    task.postCommit(true);  // <- this should call stateManager.checkpoint() internally.
    task.closeClean();  // <- this should call stateManager.close() internally
    // Those internal calls are the ones that need to be verified.

    assertEquals(Task.State.CLOSED, task.state());

    final double expectedCloseTaskMetric = 1.0;
    verifyCloseTaskMetric(expectedCloseTaskMetric, streamsMetrics, metricName);

    verify(stateManager).changelogOffsets();
    stateManagerOrdered.verify(stateManager, atLeastOnce()).close();
}
```

cadonna replied:
The test should actually look similar to this:
```java
@Test
public void shouldSuspendAndCommitBeforeCloseClean() {
    when(stateManager.changelogOffsets())
            .thenReturn(Collections.singletonMap(partition, 60L));
    final MetricName metricName = setupCloseTaskMetric();

    task = createStandbyTask();
    task.initializeIfNeeded();
    task.suspend();
    task.prepareCommit();
    task.postCommit(true);
    task.closeClean();

    assertEquals(Task.State.CLOSED, task.state());

    final double expectedCloseTaskMetric = 1.0;
    verifyCloseTaskMetric(expectedCloseTaskMetric, streamsMetrics, metricName);

    final InOrder stateManagerOrdered = inOrder(stateManager);
    stateManagerOrdered.verify(stateManager).checkpoint();
    stateManagerOrdered.verify(stateManager).close();
}
```

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StandbyTaskTest.java:
```diff
@@ -313,36 +282,29 @@ public void shouldNotFlushAndThrowOnCloseDirty() {
 
     @Test
     public void shouldSuspendAndCommitBeforeCloseClean() {
         stateManager.close();
```
Remove this call on the mock. In Mockito it does not make sense.

mdedetrich replied:
Done.

cadonna replied:
The call on the mock is still there.

mdedetrich replied:
Apologies, I think I commented/resolved the wrong thread. Just pushed the change now.

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StandbyTaskTest.java:
```diff
@@ -313,36 +282,29 @@ public void shouldNotFlushAndThrowOnCloseDirty() {
     public void shouldSuspendAndCommitBeforeCloseClean() {
         stateManager.close();
-        EasyMock.expectLastCall();
+        verify(stateManager, atLeastOnce()).close();
```
Please move this verification to after the call under test.

mdedetrich replied:
Done

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StandbyTaskTest.java:
```diff
@@ -313,36 +282,29 @@ public void shouldNotFlushAndThrowOnCloseDirty() {
     }
 
     @Test
     public void shouldSuspendAndCommitBeforeCloseClean() {
```
I think for this test you need to verify the calls on the mock with `InOrder` (see https://site.mockito.org/javadoc/current/org/mockito/InOrder.html).

mdedetrich replied:
I am unsure what needs to be verified in order? The current version of the test is 

```java
@Test
public void shouldNotThrowFromStateManagerCloseInCloseDirty() {
    doThrow(new RuntimeException("KABOOM!")).when(stateManager).close();

    task = createStandbyTask();
    task.initializeIfNeeded();

    task.suspend();
    task.closeDirty();
}
```

And from what I can tell its testing that the `RuntimeException` ie never thrown when `task.closeDirty()` is called (`task.closeDirty()` indirectly calls `stateManager.close()`). The only mock is the `doThrow` so there is nothing to really order?

I can add `verify(stateManager).close();` at the end of the test but even then there is only a single `.verify` statement.

cadonna replied:
Actually, my comment is on line 305, i.e., method `shouldSuspendAndCommitBeforeCloseClean()` and not `shouldNotThrowFromStateManagerCloseInCloseDirty()`. I think there you should verify that `stateManager.checkpoint` is called before `stateManager.close()`. But since the current test does also not check the order and this PR is about migration, you can consider this an optional request.

mdedetrich replied:
Done

cadonna commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StandbyTaskTest.java:
```diff
@@ -313,36 +282,29 @@ public void shouldNotFlushAndThrowOnCloseDirty() {
-        EasyMock.expectLastCall().once();
-        EasyMock.expect(stateManager.changelogOffsets())
-                .andReturn(Collections.singletonMap(partition, 60L));
-        EasyMock.expect(stateManager.changelogPartitions()).andReturn(Collections.singleton(partition)).anyTimes();
```
You did not migrate this line neither to a stubbing nor to a verification.

mdedetrich replied:
This was another case where I didn't know what `EasyMock.verify(stateManager)` at the end of the method was actually verifying.

I placed 

```java
when(stateManager.changelogOffsets())
    .thenReturn(Collections.singletonMap(partition, 60L));
```

back and verified `verify(stateManager).changelogOffsets();`. Note that mockito reported
```java
when(stateManager.changelogPartitions())
        .thenReturn(Collections.singleton(partition));
```
as unnecessary due to `verify(stateManager).changelogPartitions()` never actually being executed in the test run.

cadonna reviewed (commented):
@mdedetrich Thank you for the updates!

I think there are still some misunderstandings about how EasyMock works. Please revisit the relevant docs (https://easymock.org/user-guide.htm).

I did not go over all files.

mdedetrich committed:
Add verify(context).uninitialize(); in shouldRecycleStoreAndReregisterChangelog

mdedetrich committed:
Put back stateManager.registerStateStores(any(), any()) in shouldTransitToRunningAfterInitialization

mdedetrich committed:
Add stateManager.flush() to shouldAlwaysCheckpointStateIfEnforced

mdedetrich committed:
Remove unnecessary verify(stateManager, atLeastOnce()).close(); from shouldSuspendAndCommitBeforeCloseClean

mdedetrich committed:
Remove verify(stateManager, atLeastOnce()).close(); from shouldNotThrowFromStateManagerCloseInCloseDirty

mdedetrich committed:
Place back stubs in shouldSuspendAndCommitBeforeCloseClean and verify with verify(stateManager).changelogOffsets()

mdedetrich committed:
Put verify(stateManager, atLeastOnce()).close() in shouldSuspendAndCommitBeforeCloseClean

mdedetrich committed:
Put verify(stateManager).flush(); in shouldOnlyCheckpointStateWithBigAdvanceIfNotEnforced

mdedetrich commented:
@cadonna I have addressed the comments which I could. I just noticed that I have gone back and forth between removing and adding the same stubs/verifications in various without realizing due to me force pushing so I pushed each of the changes as an individual commit to better track changes. If you want me to squash the PR then let me know.

mdedetrich committed:
Add verify(stateManager).close(); to shouldNotThrowFromStateManagerCloseInCloseDirty

mdedetrich committed:
Add verify(store).close() to shouldUnregisterChangelogsDuringClose

cadonna commented:
@mdedetrich Some of our discussions on this PR are still unresolved. Did you forget to push some commits?

cadonna commented:
@mdedetrich I answered your question about `InOrder`. Sorry, I somehow missed that. 

Note that I unresolved other two comments:
https://github.com/apache/kafka/pull/12524/files#r984332259
https://github.com/apache/kafka/pull/12524/files#r984332539
Let me know if my comments are not clear.

mdedetrich committed:
Add inOrder verification for shouldSuspendAndCommitBeforeCloseClean

mdedetrich committed:
Remove verify(stateManager).close(); from shouldNotThrowFromStateManagerCloseInCloseDirty

mdedetrich commented:
@cadonna 

> https://github.com/apache/kafka/pull/12524/files#r984332259

This one is resolved now, I think I somehow missed that, apologies

> https://github.com/apache/kafka/pull/12524/files#r984332539

So this one is not clear to me, commented on it.

mdedetrich committed:
Reorder verify(context, times(2)).uninitialize(); in shouldRecycleStoreAndReregisterChangelog

clolov commented:
Hello @mdedetrich and @cadonna! Is there something still remaining for this pull request or are we ready to merge it? There are around 4 pull requests remaining of this migration and I would like to get them to a completion 😊

mdedetrich commented:
Sorry I have just been busy with other things, will look into it next week

mimaison commented:
Can this be closed now? It looks like this has been done in https://github.com/apache/kafka/commit/a33c47ea4ddc810d66b6ed17ab74e40c5b7668fb / https://github.com/apache/kafka/commit/8b72a2c72f09838fdd2e7416c98d30fe876b4078

cadonna commented:
Yes
